_posts/2022-07-15-blog-wpe-graphics-architecture.md

```diff
@@ -0,0 +1,125 @@
+---
+layout: post
+title: "WPE Graphics architecture"
+tags: [blogpost]
+author: magomez
+permalink: /blog/03-wpe-graphics-architecture.html
+---
+
+Following <a href="/blog/02-overview-of-wpe.html">the previous post in the series about <a href="https://wpewebkit.org/">WPE</a> where we talked about the <a href="https://wpewebkit.org/">WPE</a> components, this post will explain briefly the <a href="https://wpewebkit.org/">WPE</a> graphics architecture, and how the engine is able to render HTML content into the display. If you haven't read the previous entries in this blog post series about <a href="https://wpewebkit.org/">WPE</a>, we recommend you to start with the <a href="/blog/01-happy-birthday-wpe.html">first post in the series</a> for an introduction, and then come back to this.
+
+
+## DOM + CSS = RenderTree
+
+As the document is parsed, it will begin building the DOM tree and load-blocking CSS resources.  At some point, possibly before the entire DOM tree is built, it's time to draw things on the screen. The first step to render the content of a page is to perform what's called the *attachment*, which is merging the DOM tree with the CSS rules, in order to create the RenderTree. This RenderTree is a collection of <a href="https://github.com/WebKit/WebKit/blob/main/Source/WebCore/rendering/RenderObject.h">RenderObjects</a>, structured into a tree, and each of these <a href="https://github.com/WebKit/WebKit/blob/main/Source/WebCore/rendering/RenderObject.h">RenderObjects</a> represent the elements in the DOM tree that have visual output. <a href="https://github.com/WebKit/WebKit/blob/main/Source/WebCore/rendering/RenderObject.h">RenderObjects</a> have the capability to render the associated DOM tree node into a surface by using the <a href="https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/GraphicsContext.h">GraphicsContext</a> class (in the case of <a href="https://wpewebkit.org/">WPE</a>, this <a href="https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/graphics/GraphicsContext.h">GraphicsContext</a> uses <a href="https://www.cairographics.org/">cairo</a> to perform the rendering).
+
+Once the RenderTree is created, the layout is performed, ensuring that each of the <a href="https://github.com/WebKit/WebKit/blob/main/Source/WebCore/rendering/RenderObject.h">RenderObjects</a> have their proper size and position set.
+
+![](/assets/graphics-attachment.png)
```

Comment:
We could fill the space in between the square brackets to provide `alt=`/`title=` attributes automatically to images. In this case it could be:

```suggestion
![Going from source content to displayed content](/assets/graphics-attachment.png)
```